src/nvim/ex_docmd.c

```diff
@@ -8874,3 +8877,100 @@ static void ex_folddo(exarg_T *eap)
   global_exe(eap->arg);
   ml_clearmarked();        /* clear rest of the marks */
 }
+
+typedef struct {
+  Job *job;
+  Terminal *term;
+  bool exited;
+} TerminalJobData;
+
+static void ex_terminal(exarg_T *eap)
+{
+  TerminalOptions topts = TERMINAL_OPTIONS_INIT;
+  topts.force = eap->forceit;
+  topts.width = curwin->w_width;
+  topts.height = curwin->w_height;
+  topts.write_cb = term_write;
+  topts.resize_cb = term_resize;
+  topts.close_cb = term_close;
+  Terminal *term = terminal_open(topts);
+  if (!term) {
+    return;
+  }
+  TerminalJobData *data = xmalloc(sizeof(TerminalJobData));
+
+  // build argument vector from the :terminal command line
+  char cmd[256];
+  snprintf(cmd, sizeof(cmd), "%s", eap->arg);
+  JobOptions opts = JOB_OPTIONS_INIT;
+  opts.data = data;
+  opts.argv = shell_build_argv(!strcmp(cmd, "") ? NULL : cmd, NULL);
+  opts.stdout_cb = on_job_output;
+  opts.stderr_cb = on_job_output;
+  opts.exit_cb = on_job_exit;
+  opts.pty = true;
+  opts.width = curwin->w_width;
+  opts.height = curwin->w_height;
+  opts.term_name = xstrdup("xterm-256color");
+  int status;
+  Job *job = job_start(opts, &status);
+  data->term = term;
+  data->job = job;
+  data->exited = false;
+
+  char title[256];
+  snprintf(title, sizeof(title), "[%s(pid: %d)]", 
+      !strcmp(cmd, "") ? (char *)p_sh : cmd, job_pid(job));
+  terminal_set_title(term, title);
+  terminal_set_data(term, data);
+
+  if (status == 0) {
+    EMSG(e_jobtblfull);
+  } else if (status < 0) {
+    EMSG(e_jobexe);
+  }
+
+  ins_char_typebuf('i'); // focus the terminal
+}
+
+static void term_write(char *buf, size_t size, void *data)
+{
+  Job *job = ((TerminalJobData *)data)->job;
+  WBuffer *wbuf = wstream_new_buffer(xmemdup(buf, size), size, 1, free);
+  job_write(job, wbuf);
+}
+
+static void term_resize(uint16_t width, uint16_t height, void *data)
+{
+  job_resize(((TerminalJobData *)data)->job, width, height);
+}
+
+static void on_job_output(RStream *rstream, void *data, bool eof)
+{
+  if (!eof) {
+    char *ptr = rstream_read_ptr(rstream);
+    size_t len = rstream_pending(rstream);
+    terminal_receive(((TerminalJobData *)job_data(data))->term, ptr, len);
+    rbuffer_consumed(rstream_buffer(rstream), len);
+  }
+}
+
+static void on_job_exit(Job *job, void *data)
+{
+  TerminalJobData *d = data;
+  if (!d->exited) {
+    d->exited = true;
+    terminal_close(d->term, _("\r\n[Program exited, press any key to close]"));
```

Comment:
We have here a memory leak when exiting nvim.
`on_job_exit()` would be called because of `job_teardown()` and `terminal_destroy()` would only be called with an additional key input.